File: fixtures/coding/batch-ingestion-ts/src/filter.ts
/** Record filtering — sign your implementation here.
 *
 * All function signatures are fixed. Fill the function bodies.
 */

import type { IngestRecord } from "./types";

export interface FilterRule {
  readonly field: string;
  readonly operator: "eq" | "neq" | "contains" | "gt" | "lt";
  readonly value: unknown;
}

export function matchesFilterRule(record: IngestRecord, rule: FilterRule): boolean {
  /**
   * Check whether a record satisfies a single filter rule.
   *
   * Supported operators:
   *   - eq / neq: equality comparison on any field
   *   - contains: value (string) in tags OR value in string payload fields
   *   - gt / lt: timestamp comparison on ts only, rule value is ISO string
   *
   * field may be: "recordType", "ts", "payload:<key>", "tags"
   * Return true if the rule is satisfied.
   */
  throw new Error("not implemented");
}

export function applyFilters(records: IngestRecord[], rules: FilterRule[]): IngestRecord[] {
  /**
   * Return only records that satisfy ALL rules (AND logic).
   *
   * Preserve original order.
   * Empty rules list → return all records unchanged.
   */
  throw new Error("not implemented");
}

File: fixtures/coding/batch-ingestion-rust/src/filter.rs
use crate::types::{PayloadValue, Record};

/// A single filter criterion.
#[derive(Clone, Debug, PartialEq)]
pub struct FilterRule {
    pub field: String,
    pub operator: String, // eq, neq, contains, gt, lt
    pub value: PayloadValue,
}

/// Check whether one record satisfies one filter rule.
///
/// Supported operators:
/// - eq / neq: equality comparison on any supported field.
/// - contains: string in tags OR substring in string payload field.
/// - gt / lt: timestamp comparison on `ts`, with integer rule value.
///
/// Field may be: `record_type`, `ts`, `payload:<key>`, or `tags`.
pub fn matches_filter_rule(_record: &Record, _rule: &FilterRule) -> bool {
    false
}

/// Return records that satisfy ALL rules (AND logic).
///
/// Empty rules returns all records unchanged. Preserve order.
pub fn apply_filters(_records: &[Record], _rules: &[FilterRule]) -> Vec<Record> {
    Vec::new()
}
